test/indexer/test_indexer_controller.py

```diff
@@ -0,0 +1,151 @@
+import json
+import logging
+from typing import (
+    Sequence,
+)
+
+import boto3
+from chalice.app import (
+    SQSRecord,
+)
+import mock
+from moto import (
+    mock_sqs,
+    mock_sts,
+)
+
+from azul import (
+    config,
+    queues,
+)
+from azul.azulclient import (
+    AzulClient,
+)
+from azul.indexer import (
+    BundleFQID,
+)
+from azul.indexer.document import (
+    Contribution,
+)
+from azul.indexer.index_controller import (
+    IndexController,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from indexer import (
+    ForcedRefreshIndexService,
+    IndexerTestCase,
+)
+
+logger = logging.getLogger(__name__)
+
+
+# noinspection PyPep8Naming
+def setUpModule():
+    configure_test_logging(logger)
+
+
+@mock_sts
+@mock_sqs
+class IndexControllerTestCase(IndexerTestCase):
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.index_service.create_indices(self.catalog)
+        self.client = AzulClient()
+        self.controller = IndexController()
+        self.queue_manager = queues.Queues(delete=True)
+
+    def tearDown(self):
+        self.index_service.delete_indices(self.catalog)
+        super().tearDown()
+
+    def _create_mock_queues(self):
+        sqs = boto3.resource('sqs')
+        for queue_name in config.all_queue_names:
+            sqs.create_queue(QueueName=queue_name)
+
+    def generate_sqs_record(self, **kwargs):
+        event_dict = {
+            'body': json.dumps(kwargs),
+            'receiptHandle': 'ThisWasARandomString',
+            'attributes': {'ApproximateReceiveCount': 1}
+        }
+        context = 'controller_test'
+        return SQSRecord(event_dict=event_dict, context=context)
+
+    def test_contribution_action_invalid(self):
+        event = [self.generate_sqs_record(action='banana', notification='slug', catalog=self.catalog)]
+        self.assertRaises(AssertionError, self.controller.contribute, event)
+
+    def test_contribution_action_reindex(self):
+        event = [self.generate_sqs_record(action='reindex', prefix='ff')]
+        with mock.patch('azul.azulclient.AzulClient.do_remote_reindex') as mock_reindex:
+            mock_reindex.return_value = True
+            self.controller.contribute(event)
+            self.assertEqual(mock_reindex.call_count, 1)
+
+    def test_contribute_and_aggregate(self):
+        """
+        Index two bundles that make contributions to the same project.
+        Inspect that the contributions match the tallies that are returned to
+        SQS. During aggregation only the project entity is deferred due to
+        multiple contributions.
+        """
+        self._create_mock_queues()
+        event = []
+        bundles = []
+        contributions_per_bundle: Sequence[Sequence[Contribution]] = []
```

Comment:
There's a type error because one of these "Sequence"s should be a "MutableSequence".